src/core/models.py

```diff
@@ -171,19 +173,34 @@ def password_rotated(self) -> bool:
         return bool(self.relation_data.get("password-rotated", None))
 
     @property
-    def hostname(self) -> str:
+    def hostname(self) -> str | None:
         """The hostname for the unit."""
-        return socket.gethostname()
+        if not self.fqdn:
+            return None
+        return self.fqdn.split(".")[0]
 
     @property
-    def fqdn(self) -> str:
+    def fqdn(self) -> str | None:
         """The Fully Qualified Domain Name for the unit."""
-        return socket.getfqdn()
+        if not (hostnames := socket.gethostbyaddr(self.private_ip)):
+            return None
+        return hostnames[0]
```

Comment:
Can you explain why we also stop relying on `socket.getfqdn()` ? 
For the following /etc/hosts:
```
127.0.0.1	localhost
255.255.255.255	broadcasthost
::1             localhost
```

Here are the outputs I'm getting:
```
> socket.getfqdn()

nebula.local
```

And for:
```
> socket.gethostname()

nebula.local
```

I think the problem is triggered by `socket.gethostbyname` (notice the `by`) when passing `127.0.0.1` as name. As it then reads from `/etc/hosts`

Reply:
So, the flow here is the contrary right, you do *not* know about `127.0.1.1`. The question the `private_ip()` tries to answer is what is your IP. If you put:

> when passing 127.0.0.1 as name

then you already know your IP, why even bother asking to `private_ip`?

------------------------------------------------------------------------------------------------------------------

In reality, we can only discover our own IP either by two means:
1. get your hostname (readily present in `/etc/hostname`)  > resolve this name as a record
2. run `ip a` > get all the IPs > get the default route IP or the IP that has a route to your destination of interest

The (2) is most reliable; but (1) is the easiest by far.

If you want to ask what is the IP of your own hostname, then you will translate that hostname into address via DNS service in your box. That service is `systemd-resolved`. If you look into its [routing rules](https://www.freedesktop.org/software/systemd/man/latest/systemd-resolved.service.html#Protocols%20and%20Routing), you will see the first rule:

> Names for which synthetic records are generated (the local hostname, "localhost" and "localdomain", local gateway, as listed in the previous section) and addresses configured in /etc/hosts are never routed to the network and a reply is sent immediately.

So, if we have our own hostname defined in the /etc/hosts as `127.0.1.1`, this is the address you will get always.

PS.: you can see the resolving chain is configured in `resolved.conf` (not resolv.conf...) and [describes its default - which is the most used way in stock Ubuntu](https://www.freedesktop.org/software/systemd/man/latest/resolved.conf.html#Options)

-----------------------------------------------------------------------------------------------------

The most reliable way is method (2). That will really get you the IP. However, it can be tricky to figure out (1) the interface of interest; and (2) your target destination.

Therefore we have Juju. And there we can just grab the relevant IP.

Reply:
@Mehdi-Bendriss did I answer your question here?